### PR TITLE
Initial import of System.Net.Sockets.

### PR DESCRIPTION
**Pull request by pgavlin**

### CODE REVIEW

**Comment by pgavlin** [comment]
@CIPop @davidsh @SidharthNabar @vijaykota @stephentoub PTAL.

**Comment by pgavlin** [comment]
Note: there are a number of issues that need to be filed to track investigating test failures. These are currently marked with "DummySomeDescriptiveIssue"; I'll update the numbers as I file the issues.

**Comment by ellismg** [comment]
@dotnet-bot test this please

**Comment by davidfowl** [comment]
Exciting!

**Comment by Clockwork-Muse** [comment]
Hmm, there's a couple instances of `Assert.True(false, ....)`, which we're generally trying to get rid of.

(PR is too big, so I can't comment on the files directly...):
SocketTestServerAsync has
```
183  if (Interlocked.Decrement(ref _acceptRetryCount) <= 0)
184  {
185      Assert.True(false, "accept retry limit exceeded.");
186      return;
187  }
```
... which should maybe be collapsed to `Assert.InRange(Interlocked.Decrement(ref _acceptRetryCount), 1, int.MaxValue);` (or possibly throwing a regular exception, given it appears to be the only assert).

DnsEndPointTests has
```
206  public void CallbackThatShouldNotBeCalled(object sender, SocketAsyncEventArgs args)
207  {
208      Assert.True(false, "This Callback should not be called");
209  }
```
... which I've generally been replacing with "ShouldNotBeInvokedException" ([waiting in a PR, though](https://github.com/dotnet/corefx/pull/3020)).

File: src/Common/src/System/Threading/Tasks/TaskToApm.cs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+            // In the future, if performance requirements drove a need, those four 
+            // allocations could be reduced to one.  This would be achieved by having TaskWrapperAsyncResult
+            // also implement ITaskCompletionAction (and optionally IThreadPoolWorkItem).  It would need
+            // additional fields to store the AsyncCallback and an ExecutionContext.  Once configured, 
+            // it would be set into the Task as a continuation.  Its Invoke method would then be run when 
+            // the antecedent completed, and, doing all of the necessary work to flow ExecutionContext, 
+            // it would invoke the AsyncCallback.  It could also have a field on it for the antecedent, 
+            // so that the End method would have access to the completed antecedent. For related examples, 
+            // see other implementations of ITaskCompletionAction, and in particular ReadWriteTask 
+            // used in Stream.Begin/EndXx's implementation.
```
Comment (stephentoub) [inline src/Common/src/System/Threading/Tasks/TaskToApm.cs]: Nit: this comment isn't relevant outside of the implementation in mscorlib, as it's referring to mscorlib internals.

File: src/System.Net.Primitives/System.Net.Primitives.sln [hunk 2]
```diff
@@ -25,10 +25,10 @@ Global
...
 		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.Debug|Any CPU.ActiveCfg = Windows_Debug|Any CPU
 		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.Debug|Any CPU.Build.0 = Windows_Debug|Any CPU
-		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.FreeBSD_Debug|Any CPU.ActiveCfg = FreeBSD_Debug|Any CPU
-		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.FreeBSD_Debug|Any CPU.Build.0 = FreeBSD_Debug|Any CPU
-		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.FreeBSD_Release|Any CPU.ActiveCfg = FreeBSD_Release|Any CPU
-		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.FreeBSD_Release|Any CPU.Build.0 = FreeBSD_Release|Any CPU
+		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.FreeBSD_Debug|Any CPU.ActiveCfg = Windows_Release|Any CPU
+		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.FreeBSD_Debug|Any CPU.Build.0 = Windows_Release|Any CPU
+		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.FreeBSD_Release|Any CPU.ActiveCfg = Windows_Release|Any CPU
+		{8772BC91-7B55-49B9-94FA-4B1BE5BEAB55}.FreeBSD_Release|Any CPU.Build.0 = Windows_Release|Any CPU
```
Comment (stephentoub) [inline src/System.Net.Primitives/System.Net.Primitives.sln]: Nit: something strange happened here

File: src/System.Net.Sockets.Legacy/tests/PerformanceTests/project.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+  "dependencies": {
+    "System.Collections": "4.0.0-beta-*",
+    "System.Diagnostics.Debug": "4.0.10-beta-*",
+    "System.Runtime": "4.0.20-beta-*",
+    "System.Threading": "4.0.0-beta-*",
+    "System.Threading.Tasks": "4.0.0-beta-*",
+    "System.Threading.Tasks.Parallel":  "4.0.0-beta-*",
```
Comment (stephentoub) [inline src/System.Net.Sockets.Legacy/tests/PerformanceTests/project.json]: @ericstj can confirm, but I believe we should be removing the "-beta-*" from most of these (same in the other project.json files)

File: src/System.Net.Sockets.Legacy/System.Net.Sockets.Legacy.sln [hunk 4]
```diff
@@ -0,0 +1,43 @@
+﻿
```
Comment (stephentoub) [inline src/System.Net.Sockets.Legacy/System.Net.Sockets.Legacy.sln]: What's the distinction between the System.Net.Sockets.Legacy and System.Net.Sockets folders?

**Comment by stephentoub** [comment]
Congrats @pgavlin and @CIPop for landing this, both the refactoring and the Unix implementations.  Great work!

Since we already reviewed all of this in the other branch, I only skimmed through this (treating it mainly as a merge) looking for anything egregious, and it looks good.  The one thing I'd recommend verifying before merging is what the project.json files should look like in terms of version numbers.

**Comment by SidharthNabar** [comment]
Address Issue #2482

File: src/System.Net.Sockets.Legacy/System.Net.Sockets.Legacy.sln (hunk 4 above)
Comment (pgavlin) [inline src/System.Net.Sockets.Legacy/System.Net.Sockets.Legacy.sln]: @SidharthNabar can address this better than I can: IIUC it has to do with plans to add a package that provides some legacy Socket APIs that are excluded from the newer surface area (e.g. the APM APIs).

File: src/System.Net.Sockets.Legacy/tests/PerformanceTests/project.json (hunk 3 above)
Comment (ericstj) [inline src/System.Net.Sockets.Legacy/tests/PerformanceTests/project.json]: Yes, anything where we have a stable package and all you need surface-area-wise is available in that stable package you should use the stable package.

File: src/System.Net.Sockets.Legacy/System.Net.Sockets.Legacy.sln (hunk 4 above)
Comment (SidharthNabar) [inline src/System.Net.Sockets.Legacy/System.Net.Sockets.Legacy.sln]: System.Net.Sockets.Legacy is created to separate the Being/End style APM methods of types in System.Net.Sockets (e.g. Socket, TcpClient). We do not want to include these in the forward-looking version of the Sockets contract, but want to provide them for backward compatibility. So we are currently planning to provide a separate contract that will take a dependency on the System.Net.Sockets contract. For more info, see https://github.com/CIPop/corefx/pull/18

File: src/System.Net.Sockets.Legacy/tests/PerformanceTests/project.json (hunk 3 above)
Comment (pgavlin) [inline src/System.Net.Sockets.Legacy/tests/PerformanceTests/project.json]: Done.

**Comment by pgavlin** [comment]
@dotnet-bot test this please

**Comment by pgavlin** [comment]
@dotnet-bot test this please

**Comment by pgavlin** [comment]
@dotnet-bot test this please

File: src/System.Net.Sockets.Legacy/System.Net.Sockets.Legacy.sln (hunk 4 above)
Comment (CIPop) [inline src/System.Net.Sockets.Legacy/System.Net.Sockets.Legacy.sln]: Also, we're still working on naming this one. (Although APM is a legacy/deprecated pattern.)

**Comment by pgavlin** [comment]
In the interests of moving things along, I'm going to go ahead and merge this. I'll file an issue to follow-up on the CR feedback.

**Comment by CIPop** [comment]
Left some comments here and in interim PRs from the dev branch.

> pgavlin merged commit 3eae6ff into dotnet:master 3 minutes ago

Wohoo! Thanks @pgavlin! :+1:

**Comment by pgavlin** [comment]
Filed #3441 to track feedback. Thanks for the reviews :)

**Pull request merged**